src/EVM/TTY.hs

```diff
@@ -317,142 +317,165 @@ takeStep
+                & set uiVm vm'
+                & set (uiVm . cache) (view (uiVm . cache) s1)
+                & set uiStep step'
+                & set uiStepper stepper'
+            in takeStep s2 (Step 0)
+          -- step until the predicate doesn't hold
+          Just (step', (vm', stepper')) ->
+            let
+              s2 = s1
+                & set uiVm vm'
+                & set (uiVm . cache) (view (uiVm . cache) s1)
+                & set uiStep step'
+                & set uiStepper stepper'
+            in takeStep s2 (StepUntil (not . p s1))
+  _ -> pure ()
 
 backstep
   :: (?fetcher :: Fetcher
      ,?maxIter :: Maybe Integer)
-  => UiVmState -> EventM n UiVmState
-backstep s = case view uiStep s of
-  -- We're already at the first step; ignore command.
-  0 -> return s
-  -- To step backwards, we revert to the previous snapshot
-  -- and execute n - 1 `mod` snapshotInterval steps from there.
-
-  -- We keep the current cache so we don't have to redo
-  -- any blocking queries, and also the memory view.
-  n ->
-    let
-      (step, (vm, stepper)) = fromJust $ lookupLT n (view uiSnapshots s)
-      s1 = s
-        & set uiVm vm
-        & set (uiVm . cache) (view (uiVm . cache) s)
-        & set uiStep step
-        & set uiStepper stepper
-      stepsToTake = n - step - 1
+  => UiVmState -> IO UiVmState
+backstep s =
+  case view uiStep s of
+    -- We're already at the first step; ignore command.
+    0 -> pure s
+    -- To step backwards, we revert to the previous snapshot
+    -- and execute n - 1 `mod` snapshotInterval steps from there.
+
+    -- We keep the current cache so we don't have to redo
+    -- any blocking queries, and also the memory view.
+    n ->
+      let
+        (step, (vm, stepper)) = fromJust $ lookupLT n (view uiSnapshots s)
+        s1 = s
+          & set uiVm vm
+          & set (uiVm . cache) (view (uiVm . cache) s)
+          & set uiStep step
+          & set uiStepper stepper
+        stepsToTake = n - step - 1
 
-    in
-      liftIO $ runStateT (interpret (Step stepsToTake) stepper) s1 >>= \case
-        (Continue steps, ui') -> return $ ui' & set uiStepper steps
-        _ -> error "unexpected end"
+      in
+        runStateT (interpret (Step stepsToTake) stepper) s1 >>= \case
+          (Continue steps, ui') -> pure $ ui' & set uiStepper steps
+          _ -> error "unexpected end"
 
 appEvent
   :: (?fetcher::Fetcher, ?maxIter :: Maybe Integer) =>
-  UiState ->
   BrickEvent Name e ->
-  EventM Name (Next UiState)
+  EventM Name UiState ()
 
 -- Contracts: Down - list down
-appEvent (ViewContracts s) (VtyEvent e@(V.EvKey V.KDown [])) = do
-  s' <- handleEventLensed s
-    browserContractList
-    handleListEvent
-    e
-  continue (ViewContracts s')
+appEvent (VtyEvent e@(V.EvKey V.KDown [])) = get >>= \case
+  ViewContracts _s -> do
+    Brick.zoom
+      (_ViewContracts . browserContractList)
+      (handleListEvent e)
+    pure ()
+  ViewVm s ->
+    if view uiShowMemory s then
```

Comment:
I think we can just ignore the `down` and `up` keys for the memory view. It's already scrollable with the scrollwheel

Reply:
done